crypto/fipsmodule/bn/exponentiation.c

```diff
@@ -1207,6 +1207,94 @@ int BN_mod_exp_mont_consttime(BIGNUM *rr, const BIGNUM *a, const BIGNUM *p,
   return ret;
 }
 
+
+/*
+ * This is a variant of modular exponentiation optimization that does
+ * parallel 2-primes exponentiation using 256-bit (AVX512VL) AVX512_IFMA ISA
+ * in 52-bit binary redundant representation.
+ * If such instructions are not available, or input data size is not supported,
+ * it falls back to two BN_mod_exp_mont_consttime() calls.
+ */
+int BN_mod_exp_mont_consttime_x2(BIGNUM *rr1, const BIGNUM *a1, const BIGNUM *p1,
+                                 const BIGNUM *m1, const BN_MONT_CTX *in_mont1,
+                                 BIGNUM *rr2, const BIGNUM *a2, const BIGNUM *p2,
+                                 const BIGNUM *m2, const BN_MONT_CTX *in_mont2,
+                                 BN_CTX *ctx)
+{
+    int ret = 0;
+
+#ifdef RSAZ_ENABLED
+    BN_MONT_CTX *mont1 = NULL;
+    BN_MONT_CTX *mont2 = NULL;
+
+    if (ossl_rsaz_avx512ifma_eligible() &&
+        (((a1->width == 16) && (p1->width == 16) && (BN_num_bits(m1) == 1024) &&
+          (a2->width == 16) && (p2->width == 16) && (BN_num_bits(m2) == 1024)) ||
+         ((a1->width == 24) && (p1->width == 24) && (BN_num_bits(m1) == 1536) &&
+          (a2->width == 24) && (p2->width == 24) && (BN_num_bits(m2) == 1536)) ||
+         ((a1->width == 32) && (p1->width == 32) && (BN_num_bits(m1) == 2048) &&
+          (a2->width == 32) && (p2->width == 32) && (BN_num_bits(m2) == 2048)))) {
+
+        int widthn = a1->width;
+        /* Modulus bits of |m1| and |m2| are equal */
+        int mod_bits = BN_num_bits(m1);
+
+        if (!bn_wexpand(rr1, widthn))
+            goto err;
+        if (!bn_wexpand(rr2, widthn))
+            goto err;
+
+        /*  Ensure that montgomery contexts are initialized */
+        if (in_mont1 == NULL) {
+          if ((mont1 = BN_MONT_CTX_new()) == NULL)
+              goto err;
+          if (!BN_MONT_CTX_set(mont1, m1, ctx))
+              goto err;
+          in_mont1 = mont1;
+        }
+        if (in_mont2 == NULL) {
+          if ((mont2 = BN_MONT_CTX_new()) == NULL)
+              goto err;
+          if (!BN_MONT_CTX_set(mont2, m2, ctx))
+              goto err;
+          in_mont2 = mont2;
+        }
+
+        ret = ossl_rsaz_mod_exp_avx512_x2(rr1->d, a1->d, p1->d, m1->d,
+                                          in_mont1->RR.d, in_mont1->n0[0],
+                                          rr2->d, a2->d, p2->d, m2->d,
+                                          in_mont2->RR.d, in_mont2->n0[0],
+                                          mod_bits);
+
+        rr1->width = widthn;
+        rr1->neg = 0;
+        bn_set_minimal_width(rr1);
+
+        rr2->width = widthn;
+        rr2->neg = 0;
+        bn_set_minimal_width(rr2);
+
+	goto err;
```

Comment:
If I remove this `goto` and AVX512 is available, we compute the exponent twice; once with avx512 in parallel and a second time serially with the original implementations below. When I step through it with gdb or leave `printf`s in to show the results—contained in `rr1` and `rr2` respectively—in the test cases that fail `rr1->d` does not contain the same result coming out of the avx512 computation as it does coming out of the serial implementation. `rr2->d` is consistently the same.

Reply:
Further notes. Here I've swapped the set of args, 2 for 1 and vice versa. 

```c
        ret = ossl_rsaz_mod_exp_avx512_x2(rr2->d, a2->d, p2->d, m2->d,
                                          in_mont2->RR.d, in_mont2->n0[0],
                                          rr1->d, a1->d, p1->d, m1->d,
                                          in_mont1->RR.d, in_mont1->n0[0],
                                          mod_bits);

        rr1->width = widthn;
        rr1->neg = 0;
        bn_set_minimal_width(rr1);

        rr2->width = widthn;
        rr2->neg = 0;
        bn_set_minimal_width(rr2);

	printf("a 1 %lu\n", *rr1->d);
	printf("a 2 %lu\n", *rr2->d);
    }
#endif

    /* rr1 = a1^p1 mod m1 */
    ret = BN_mod_exp_mont_consttime(rr1, a1, p1, m1, ctx, in_mont1);
    /* rr2 = a2^p2 mod m2 */
    ret &= BN_mod_exp_mont_consttime(rr2, a2, p2, m2, ctx, in_mont2);

    printf("b 1 %lu\n", *rr1->d);
    printf("b 2 %lu\n", *rr2->d);
```

In this case, `a 1` consistently equals `b1`, but they are the second set of arguments given to `ossl_rsaz_mod_exp_avx512_x2`. Something about how this function is computing its first set is inconsistent with how the serial / single implementation is computing it, but I lack the understanding of what its doing to understand why.